### PR TITLE
Fix a bug where FAST_WITH_THRESHOLD executions weren't counted

### PR DESCRIPTION
**Pull request by WildFireFlum**
Added a new metric for FAST_WITH_THRESHOLD executions
Added apollo support for the new metric
Extracted a function for execution related metric updates
Added an opt-in makefile option to skip format when building via the
SKIP_FORMAT enviroment variable

### CODE REVIEW

**Comment by evdzhurov** [comment]
Are there any changes compared to the previous PR or this is strictly the split code?

**Comment by WildFireFlum** [comment]
> Are there any changes compared to the previous PR or this is strictly the split code?

No new changes.

**Review by yontyon — APPROVED** [review]
LGTM!
Thanks

**Comment by evdzhurov** [comment]
My main concern is that this PR changes the logic of how the commit path metrics are gathered, including those used by _wait_for_consensus_path_to_be_prevalent. And I can already see from the test runs that there is instability introduced in some cases. You can either consider to address the instabilities of _wait_for_consensus_path_to_be_prevalent or limit the rework to metrics needed only for the new test_fast_to_fast_with_threshold_path .

**Comment by yontyon** [comment]
@evdzhurov , this should be merged ASAP. Please review it once you have time (even if it merged) and provide your imput

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -5840,6 +5814,28 @@ void ReplicaImp::tryToGoToNextView() {
 
 IncomingMsgsStorage &ReplicaImp::getIncomingMsgsStorage() { return *msgsCommunicator_->getIncomingMsgsStorage(); }
 
+void ReplicaImp::startSlowPath(SeqNumInfo &seqNumInfo) {
+  seqNumInfo.startSlowPath();
+  metric_total_slowPath_++;
+}
+void ReplicaImp::updateCommitMetrics(const CommitPath &commitPath) {
+  LOG_INFO(CNSUS, "Updating metric" << commitPath);
```
Comment (evdzhurov) [inline bftengine/src/bftengine/ReplicaImp.cpp:5822]: I'm not sure there needs to be an INFO level log every time you update a metric. Is this necessary?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -2208,7 +2202,8 @@ void ReplicaImp::onFastPathCommitCombinedSigSucceeded(SeqNum seqNumber,
   auto span = concordUtils::startChildSpanFromContext(fcp->spanContext<std::remove_pointer<decltype(fcp)>::type>(),
                                                       "bft_execute_committed_reqs");
 
-  metric_total_committed_sn_++;
+  auto commitPath = seqNumInfo.getFastPathSelfPartialCommitProofMsg()->commitPath();
```
Comment (evdzhurov) [inline bftengine/src/bftengine/ReplicaImp.cpp:2205]: The commit path is already known as a function argument of onFastPathCommitCombinedSigSucceeded.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -2273,7 +2268,8 @@ void ReplicaImp::onFastPathCommitVerifyCombinedSigResult(SeqNum seqNumber,
...
```
Comment (evdzhurov) [inline bftengine/src/bftengine/ReplicaImp.cpp:2271]: The commit path is already known as a function argument of onFastPathCommitVerifyCombinedSigResult

File: tests/apollo/test_skvbc_commit_path.py [hunk 4]
```diff
@@ -58,20 +87,19 @@ async def test_fast_path_is_default(self, bft_network, tracker, exchange_keys=Tr
...
-        await bft_network.wait_for_fast_path_to_be_prevalent(
-            run_ops=lambda: skvbc.run_concurrent_ops(num_ops=NUM_OPS, write_weight=1), threshold=NUM_OPS)
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network, tracker)
 
+        await bft_network.wait_for_consensus_path(path_type=ConsensusPathType.OPTIMISTIC_FAST,
+                                                  run_ops=lambda: self.send_kvs_sequentially(skvbc, op_count),
+                                                  threshold=op_count)
 
-    @unittest.skip("This is a transition covered in test_commit_path_transitions and is kept as a manual testing option.")
+    # @unittest.skip("This is a transition covered in test_commit_path_transitions and is kept as a manual testing option.")
```
Comment (evdzhurov) [inline tests/apollo/test_skvbc_commit_path.py:98]: Did you mean to re-enable this test or you simply uncommented to test it?

File: tests/apollo/test_skvbc_commit_path.py [hunk 5]
```diff
@@ -85,24 +113,58 @@ async def test_fast_to_slow_path(self, bft_network, tracker):
 
         Finally the decorator verifies the KV execution.
         """
-
+        op_count = int(self.EVALUATION_PERIOD_SEQUENCES * 0.1)
```
Comment (evdzhurov) [inline tests/apollo/test_skvbc_commit_path.py:116]: Why did you choose the 0.1 factor here? Maybe a comment in code would help.

File: tests/apollo/test_skvbc_commit_path.py [hunk 6]
```diff
@@ -85,24 +113,58 @@ async def test_fast_to_slow_path(self, bft_network, tracker):
...
+        crash_targets = random.sample(bft_network.all_replicas(without={primary_num}), bft_network.config.c)
+        bft_network.stop_replicas(crash_targets)
+
+        await bft_network.wait_for_consensus_path(path_type=ConsensusPathType.FAST_WITH_THRESHOLD,
+                                                  run_ops=lambda: self.send_kvs_sequentially(skvbc, ops_for_transition),
+                                                  threshold=1)
+
+        slow_commits = await bft_network.commit_count_by_path(ConsensusPathType.SLOW)
+        assert slow_commits > 0, f"Invalid slow commit count on transition from fast path to fast with threshold, " \
+                                 f"slow count {slow_commits}"
```
Comment (evdzhurov) [inline tests/apollo/test_skvbc_commit_path.py:164]: Is the purpose of the assert here to make sure that transitions between fast optimistic to fast with threshold involve commits on the slow path?

File: tests/apollo/test_skvbc_commit_path.py [hunk 7]
```diff
@@ -185,23 +229,24 @@ async def test_fast_path_after_view_change(self, bft_network, tracker):
...
+            run_ops=lambda: self.send_kvs_sequentially(skvbc, num_ops),
+            threshold=num_ops)
 
         # Stop the primary
         bft_network.stop_replica(0)
 
         # Send a write request to trigger a view change
         with trio.move_on_after(seconds=3):
-            await skvbc.send_write_kv_set()
+            await skvbc.run_concurrent_ops(1, write_weight=1)
```
Comment (evdzhurov) [inline tests/apollo/test_skvbc_commit_path.py:246]: Why did you need to change it to run_concurrent_ops?

**Review by evdzhurov — APPROVED** [review]
LGTM with minor comments.

File: tests/apollo/test_skvbc_commit_path.py (hunk 6 above)
Comment (WildFireFlum) [reply tests/apollo/test_skvbc_commit_path.py:164]: Yes, I'll change the assertion's text

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (WildFireFlum) [reply bftengine/src/bftengine/ReplicaImp.cpp:5822]: Removed.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 2 above)
Comment (WildFireFlum) [reply bftengine/src/bftengine/ReplicaImp.cpp:2205]: Changed to use the passed parameter

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 3 above)
Comment (WildFireFlum) [reply bftengine/src/bftengine/ReplicaImp.cpp:2271]: Changed to use the parameter

File: tests/apollo/test_skvbc_commit_path.py (hunk 4 above)
Comment (WildFireFlum) [reply tests/apollo/test_skvbc_commit_path.py:98]: Uncommented to test it

File: tests/apollo/test_skvbc_commit_path.py (hunk 5 above)
Comment (WildFireFlum) [reply tests/apollo/test_skvbc_commit_path.py:116]: Added comment
